gax-java/gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonLoggingInterceptor.java

```diff
@@ -0,0 +1,197 @@
+      HttpJsonChannel next) {
+
+    String endpoint = ((ManagedHttpJsonChannel) next).getEndpoint();
+
+    return new SimpleForwardingHttpJsonClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+
+      LogData.Builder logDataBuilder = LogData.builder();
+
+      @Override
+      public void start(
+          HttpJsonClientCall.Listener<RespT> responseListener, HttpJsonMetadata headers) {
+
+        logRequestInfo(method, endpoint, logDataBuilder);
+        recordRequestHeaders(headers, logDataBuilder);
+
+        Listener<RespT> forwardingResponseListener =
+            new SimpleForwardingHttpJsonClientCallListener<RespT>(responseListener) {
+
+              @Override
+              public void onHeaders(HttpJsonMetadata responseHeaders) {
+                recordResponseHeaders(responseHeaders, logDataBuilder);
+                super.onHeaders(responseHeaders);
+              }
+
+              @Override
+              public void onMessage(RespT message) {
+                recordResponsePayload(message, logDataBuilder);
+                super.onMessage(message);
+              }
+
+              @Override
+              public void onClose(int statusCode, HttpJsonMetadata trailers) {
+                try {
+                  logResponse(statusCode, logDataBuilder);
+                } finally {
+                  logDataBuilder = null; // release resource
+                }
+                super.onClose(statusCode, trailers);
+              }
+            };
+        super.start(forwardingResponseListener, headers);
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        logRequestDetails(message, logDataBuilder);
+        super.sendMessage(message);
+      }
+    };
+  }
+
+  // Helper methods for logging,
+  // some duplications with grpc equivalent to avoid exposing as public method
+  private <ReqT, RespT> void logRequestInfo(
+      ApiMethodDescriptor<ReqT, RespT> method, String endpoint, LogData.Builder logDataBuilder) {
+    try {
+      if (logger.isInfoEnabled()) {
+        logDataBuilder
+            .rpcName(method.getFullMethodName())
+            .httpMethod(method.getHttpMethod())
+            .httpUrl(endpoint);
+
+        if (!logger.isDebugEnabled()) {
+          LoggingUtils.logWithMDC(
+              logger, Level.INFO, logDataBuilder.build().toMapRequest(), "Sending HTTP request");
+        }
+      }
+    } catch (Exception e) {
+      logger.error("Error logging request info (and headers)", e);
+    }
+  }
+
+  private void recordRequestHeaders(HttpJsonMetadata headers, LogData.Builder logDataBuilder) {
+    try {
+      if (logger.isDebugEnabled()) {
+        JsonObject requestHeaders = new JsonObject();
+        headers
+            .getHeaders()
+            .forEach((key, value) -> requestHeaders.addProperty(key, value.toString()));
+        logDataBuilder.requestHeaders(gson.toJson(requestHeaders));
+        logDataBuilder.requestHeaders(gson.toJson(requestHeaders));
```

Comment:
Why are we calling this twice?

Reply:
That's a mistake. Thanks for catching this!